### PR TITLE
Trivial object support for embind

### PR DESCRIPTION
**Pull request by crazyjul**
Embind currently supports 2 mode for object exposure : value object and referenced object. The system is great for most types, but there a category of types that does not fit those 2 categories, mostly math types ( vector, matrix ).

The value object mode is too expensive as instance needs to be converted recursively every time it pass the c++/js boundary. On the other end, the referenced object needs to be deleted explicitly though delete. This is not convenient for such types.

This PR is introducing the support of trivial object. By C++ definition, a trivial object is one that does not have non-trivial constructor, copy-constructor and destructor. It can be though as POD object, or raw object. Those C++ classes can be managed as a memory buffer.

The idea is to copy the data of an instance into a separate ArrayBuffer.  So the object is kept byte-wise by the JS object. Every time the object needs to be modified, it is copied back in the heap before calling the C++ function. As the ArrayBuffer is separated from the heap one, the object can be garbage collected.

The work is not completely finished yet, but I would like to have early comments and advices.

The next major feature is to use property to access the buffer directly, e.g. vector.X would not need to go back to the heap.

Any interests, advices, comments?

### CODE REVIEW

**Comment by kripken** [comment]
cc @chadaustin (no rush of course, whenever you can take a look)

@crazyjul: note that he's probably not going to be able to get to this for a while.

**Comment by crazyjul** [comment]
Yeah, I know, I follow the discussions on every issues :-( . No hurry for this one

File: src/embind/embind.js [hunk 1]
```diff
@@ -2147,6 +2147,363 @@ var LibraryEmbind = {
...
+  $RegisteredTrivialObject_toWireType__deps:[
+      '$RegisteredTrivialObject_copyAndFreePtr', '$RegisteredTrivialObject_freePtr'],
+  $RegisteredTrivialObject_toWireType: function(destructors, object) {
+
+      //:TODO: verify alignment
+    var local_heap = new Uint32Array( object.$$.buffer );
+    var ptr = _malloc( this.size );
+    var heap = HEAPU32;
+
+    for(i=0; i< this.size; i+=4){
```
Comment (chadaustin) [inline src/embind/embind.js]: If you don't mind, could you update the code to match the rest of the file in terms of whitespace?

**Comment by chadaustin** [comment]
Hello, I am back from leave now.

Interesting idea!  Would the ArrayBuffer ever be modified once it's copied?  

Seems like a cool idea.  Adds a lot of code, unfortunately.  Is there any way to refactor some of the commonality with normal classes?  In your example, would e.setFloat(10) result in two copies?  One into the heap and then one back into the ArrayBuffer?  It seems a little dangerous to have performance characteristics like that hidden behind innocuous code, BUT it's nice to have the option to avoid calling .delete().  :)

I would support this functionality, especially if you can find some good way to share some of the code between the class implementations, if possible.

File: src/embind/embind.js (hunk 1 above)
Comment (crazyjul) [inline src/embind/embind.js]: I'll do it with class code merging.

**Comment by crazyjul** [comment]
I thought a lot about code sharing, but this was more a proof of concept, easier to see the idea with separate code. I'll merge it with the class code.

For easier sharing, I think of changing the ```toWire()``` code. The code is currently making a copy, through copy constructor. We could defer that copy to the js code, passing an additional clone function. This way, class code would copy the object, but trivial object would not need to. As the clone function is using pointer in and out, that would not be too costly.

For the setFloat(), the copy back and forth can be costly, but heap to ArrayBuffer is native, should be fast for small objects.  but I saw experimental function that could allow merging back natively. I agree the copy back to Heap is currently slow. Also, const parameters are not copied back. 

The properties will access the ArrayBuffer directly, using the offset of the field in the object. I'm currently finishing that part. The initial idea for this code was vectors and quaternions. accessing field is most of what is needed.

If all this is ok, I'll continue working on this.

**Comment by chadaustin** [comment]
Sounds fine to me.  Thanks!

**Comment by abcthomas** [comment]
Is this work complete ?

**Comment by crazyjul** [comment]
Hi Thomas,

No further work as been done for now, but I still plan to work on it soon.
Do you need it, or is it to close the MR?

**Comment by abcthomas** [comment]
I'm pretty much in the same boat as you, trying to develop a large interface that utilises vectors and matrices, I had some simple vector value types but I've just discovered that those can't actually be passed directly back into c++, so yeah just very interested in this.

**Comment by crazyjul** [comment]
Ok, I'll work on this in the next few days. Expect a new version for
monday. It's always easier to get motivated if someone else needs it :-)

**Comment by kripken** [comment]
This PR is quite old, closing. Feel free to reopen if still relevant.

**Pull request closed without merging**